x-pack/plugins/index_lifecycle_management/public/store/selectors/lifecycle.js

```diff
@@ -0,0 +1,234 @@
+  STRUCTURE_INDEX_TEMPLATE,
+  STRUCTURE_CONFIGURATION,
+  STRUCTURE_PRIMARY_NODES,
+  STRUCTURE_REPLICAS,
+  STRUCTURE_TEMPLATE_SELECTION,
+  STRUCTURE_TEMPLATE_NAME,
+  STRUCTURE_POLICY_NAME,
+  STRUCTURE_POLICY_CONFIGURATION,
+  STRUCTURE_INDEX_NAME,
+  STRUCTURE_ALIAS_NAME,
+  ERROR_STRUCTURE,
+  PHASE_ATTRIBUTES_THAT_ARE_NUMBERS,
+  PHASE_PRIMARY_SHARD_COUNT,
+  PHASE_SHRINK_ENABLED,
+  STRUCTURE_REVIEW,
+  PHASE_FORCE_MERGE_ENABLED,
+  PHASE_FORCE_MERGE_SEGMENTS
+} from '../constants';
+import {
+  getPhase,
+  getPhases,
+  phaseToES,
+  getSelectedPolicyName,
+  getSelectedIndexTemplateName,
+  isNumber,
+  getSelectedPrimaryShardCount,
+  getSelectedReplicaCount,
+  getSaveAsNewPolicy,
+  getSelectedOriginalPolicyName,
+  getBootstrapEnabled,
+  getIndexName,
+  getAliasName,
+} from '.';
+
+export const validatePhase = (type, phase) => {
+  const errors = {};
+
+  if (!phase[PHASE_ENABLED]) {
+    return errors;
+  }
+
+  if (phase[PHASE_ROLLOVER_ENABLED]) {
+    if (
+      !isNumber(phase[PHASE_ROLLOVER_MAX_AGE]) &&
+      !isNumber(phase[PHASE_ROLLOVER_MAX_SIZE_STORED])
+    ) {
+      errors[PHASE_ROLLOVER_MAX_AGE] = [
+        'A maximum age is required'
+      ];
+      errors[PHASE_ROLLOVER_MAX_SIZE_STORED] = [
+        'A maximum index size is required'
+      ];
+    }
+  }
+
+  for (const numberedAttribute of PHASE_ATTRIBUTES_THAT_ARE_NUMBERS) {
+    if (phase.hasOwnProperty(numberedAttribute) && phase[numberedAttribute] !== '') {
+      // If shrink is disabled, there is no need to validate this
+      if (numberedAttribute === PHASE_PRIMARY_SHARD_COUNT && !phase[PHASE_SHRINK_ENABLED]) {
+        continue;
+      }
+      if (!isNumber(phase[numberedAttribute])) {
+        errors[numberedAttribute] = ['A number is required'];
+      }
+      else if (phase[numberedAttribute] < 0) {
+        errors[numberedAttribute] = ['Only positive numbers are allowed'];
+      }
+      else if (numberedAttribute === PHASE_PRIMARY_SHARD_COUNT && phase[numberedAttribute] < 1) {
+        errors[numberedAttribute] = ['Only positive numbers are allowed'];
+      }
+    }
+  }
+
+  if (phase[PHASE_SHRINK_ENABLED]) {
+    if (!isNumber(phase[PHASE_PRIMARY_SHARD_COUNT])) {
+      errors[PHASE_PRIMARY_SHARD_COUNT] = ['A number is required.'];
+    }
+    else if (phase[PHASE_PRIMARY_SHARD_COUNT] < 1) {
+      errors[PHASE_PRIMARY_SHARD_COUNT] = ['Only positive numbers above 0 are allowed.'];
+    }
+  }
+
+  if (phase[PHASE_FORCE_MERGE_ENABLED]) {
+    if (!isNumber(phase[PHASE_FORCE_MERGE_SEGMENTS])) {
+      errors[PHASE_FORCE_MERGE_SEGMENTS] = ['A number is required.'];
+    }
+    else if (phase[PHASE_FORCE_MERGE_SEGMENTS] < 1) {
+      errors[PHASE_FORCE_MERGE_SEGMENTS] = ['Only positive numbers above 0 are allowed.'];
+    }
+  }
+
+  return errors;
+};
+
+export const validateLifecycle = state => {
+  // This method of deep copy does not always work but it should be fine here
+  const errors = JSON.parse(JSON.stringify(ERROR_STRUCTURE));
```

Comment:
It might not be worth it right now, but I think it would great to see if we could come up with a simpler way of storing validation errors. I'm not sure what the benefit of this deep structure is. I find it very difficult to parse paths like this:

```js
errors[STRUCTURE_INDEX_TEMPLATE][STRUCTURE_TEMPLATE_SELECTION][STRUCTURE_INDEX_NAME]
```

I think a flat object would be easier for me to parse and it would be more robust because it would be less coupled to the flow of our wizard. That would make it easier to change up the wizard flow by moving around its sections and form elements without requiring changes to the validation.